sympy/sets/conditionset.py

```diff
@@ -191,8 +191,12 @@ def __new__(cls, sym, condition, base_set=S.UniversalSet):
 
     @property
     def free_symbols(self):
-        s, c, b = self.args
-        return (c.free_symbols - s.free_symbols) | b.free_symbols
+        cond_syms = self.condition.free_symbols - self.sym.free_symbols
+        return cond_syms | self.base_set.free_symbols
+
+    @property
+    def bound_symbols(self):
+        return self.sym.free_symbols
```

Comment:
Is the invariant `self.atoms(Symbol) == self.free_symbols | self.bound_symbols`? If so, then should this return something like `self.condition.bound_symbols | {self.sym} | self.base_set.bound_symbols`?

Reply:
That's what I was wondering too! I've just added a few tests to illustrate. For e.g. `Sum` the `bound_symbols` seem to be determined "recursively", i.e. it's their union.

Reply:
I must confess that I find it somewhat difficult to untangle `dummy_eq`, `as_dummy`, etc. due to very - ahem - terse docstrings. In particular I'm still not sure I 100% understand 
- what `dummy_eq` is supposed to do
- what `dummy_eq` does exactly.

I found some answers in an older discussion in the issues/PRs though, but that's hardly "documentation". (It seems to do two things at a time a) matching `Dummy`s to `Symbol`s and b) comparing expressions with bound variables.)

Also I find it misleading that in the docstrings bound variables are called "dummy variables" without warning or explanation that they're not to be confused with `Dummy` variables.

Reply:
See https://github.com/sympy/sympy/pull/19261#issuecomment-626401765

@smichr and I recently discussed exactly this and I think the conclusion was that `dummy_eq` should be changed to something basically (or literally?) like `a.as_dummy()==b.as_dummy()` so that its sole purpose is comparing modulo the identity of bound symbols.

Making that change just didn't happen yet...

Reply:
I think it should be:
```python
    def bound_symbols(self):
        return self.sym.free_symbols | self.condition.bound_symbols | self.base_set.bound_symbols
```
Note that sym might not be a single symbol e.g.:
```julia
In [34]: ConditionSet((x, y), x<=y, Reals**2)                                                                                                  
Out[34]: 
⎧                   2          ⎫
⎨(x, y) | (x, y) ∊ ℝ  ∧ (x ≤ y)⎬
⎩                              ⎭
```
I think `ConditionSet` needs some work in the multivariate case analogous to what was done with `Lambda` and `ImageSet`.

Reply:
> ```python
>     def bound_symbols(self):
>         return self.sym.free_symbols | self.condition.bound_symbols | self.base_set.bound_symbols
> ```
Hmm. `Eq` and probably other relationals don't have `.bound_symbols`. In fact almost none of the classes have. So I'd have to filter the expression trees of `condition` and `base_set` to actually find bound symbols? Or better yet add it to `Basic` similarly to `free_symbols`?
https://github.com/sympy/sympy/blob/32c28ebf80b6056b07c2a0d8bb4dc249741b86a4/sympy/core/basic.py#L526-L539

Reply:
Yes, probably it should be on Basic along with free symbols. The default implementation can just be:
```python
    @property
    def bound_symbols(self):
        return set().union(*(arg.bound_symbols for arg in self.args))
```

Reply:
Then any class that wants to add to the bound symbols can do e.g.:
```python
    @property
    def bound_symbols(self):
        return super().bound_symbols | self.sym.free_symbols
```
I don't think `super()` can be used with a `property` though... Something like `Basic.bound_symbols.fget(self) | self.sym.free_symbols`.

Reply:
> so that its sole purpose is comparing modulo the identity of bound symbols.

agree. And if functions would never use Dummy except internally, picking a canonical, non-clashing symbol instead, that would make testing without strings easier. e.g. you are able to write the tests in this PR using `Symbol(_0)` because that symbol is not a Dummy.